pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import os
+
+from pootle.core.url_helpers import split_pootle_path
+from pootle_store.models import Store
+
+
+def split_vfolder_location(path):
+    (lang_code, proj_code,
+     dir_path, filename) = split_pootle_path(path)
+    if lang_code == "{LANG}":
+        lang_code = None
+    if proj_code == "{PROJ}":
+        proj_code = None
+    return lang_code, proj_code, dir_path, filename
+
+
+class VirtualFolderStoreMatcher(object):
+
+    def __init__(self, vf):
+        self.vf = vf
+
+    def add_related_stores(self):
+        existing_stores = self.vf.stores.all()
+        related_stores = self.find_related_stores()
+        for store in related_stores:
+            if store not in existing_stores:
+                self.vf.stores.add(store)
+
+    def add_store_if_related(self, store):
+        if store in self.find_related_stores():
```

Comment:
I find this ineficient. First I would check if it is already in `self.vf.stores.all()` and if it is not in there I wouldn't user `find_related_stores` to find out if store matches any rule since it hits the DB. There should be a faster way of matching the store's pootle_path against all the location+filter_rules.